### PR TITLE
Introduce init and modules. Close #406.

### LINKED ISSUE #406: Introduce init and modules.
We don't have an elegant way to call any actions, subscribe to global events or kickstart ourselves when our app loads. We can call actions after we return from `app()` but packages like the router, a mouse / keyboard interop interface, etc., could be exposed as **modules** and for their initialization needs, we can introduce a new **init** function. 🎉 

> By the way, modules are not like mixins as they would be scoped to a state & action slice, that you can't opt out from.

```jsx
import { whopper } from "./whooper"
import { mouse } from "./mouse"

import state from "./myState"
import actions from "./myActions"
import view from "./myView"

app({
  init(state, actions) {
    // Subscribe to global events, start timers, fetch resources & more!
  },
  state,
  actions,
  view,
  modules: { whopper, mouse }
})
```

The modules `whopper` and `mouse` of this example can definitely expose their own `init` function. They would be called after the top-level `init` with a slice of the state and actions, e.g., `state["whopper"]` and `actions["whopper"]`,  just how it works with actions.

In @Zaceno's words, what's the difference between modules and mixins and how do we explain their overlap with HOAs? 

> This actually gets at the core of the problem I saw with mixins (and why I wanted them renamed from plugins as they once were called): they basically served _two_ purposes: either to A) augment hyperapp’s features, or B) to modularize your app. **Now**, modules is only for (B), and we have HOA for (A).

/cc @lukejacksonn @zaceno @andyrj @Swizz

### PR DESCRIPTION
**Pull request by JorgeBucaran**
## tl;dr

- Introduce `init` function (essentially the original `events.load`).
- Introduce `modules`, similar to the old mixins, but without all their issues. See #406 for a full explanation.
  - Modules can have an init function too.
  - Modules can have nested modules.

### `init(state, actions) { ... }`

We don't have an elegant way to call any actions, subscribe to global events or kickstart ourselves when our app loads.

```jsx
app({
  init(state, actions) {
    // Subscribe to global events, start timers, fetch resources & more!
  }
})
```

### `modules: { foo, bar, baz  }`

We also need a way to encapsulate code without all the issues mixins had.

```jsx
app({
  modules: { router, whopper, bazinga }
})
```

Desugars into:

```jsx
app({
  init(state, actions) {
    // Close enough. Actual order is non-deterministic. 
    // Also, router, whooper and bazinga `init` function runs BEFORE the top level `init`.
    // This is consistent with sub-modules. Deepest first. Inside-out.

    router.init(state.router, actions.router)
    whopper.init(state.whopper, actions.whopper)
    bazinga.init(state.bazinga, actions.bazinga)
  },
  state: { 
    router: router.state,
    whopper: whopper.state,
    bazinga: bazinga.state,
  },
  actions: { 
    router: router.actions,
    whopper: whopper.actions,
    bazinga: bazinga.actions
  }
})
```

Quoting @zaceno on the difference between mixins and modules and how they fit together with HOAs.

> This actually gets at the core of the problem I saw with mixins (and why I wanted them renamed from plugins as they once were called): they basically served _two_ purposes: either to A) augment hyperapp’s features, or B) to modularize your app. **Now**, modules is only for (B), and we have HOA for (A).

Close #406, #404.

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -1,21 +1,21 @@
 import { h } from "./h"
 
-var lifecycleCallbackStack = []
+var callbackStack = []
```
Comment (pspeter3) [inline src/app.js]: Just curious, why stack as opposed to queue?

File: src/app.js [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 
 export function app(props) {
+  if (typeof props === "function") {
```
Comment (pspeter3) [inline src/app.js]: This is what provides HOA syntax right?

File: src/app.js [hunk 3]
```diff
@@ -25,107 +25,92 @@ export function app(props) {
...
+    while ((cb = callbackStack.pop())) cb()
+  }
+
+  function setup(module, state, actions) {
+    copy(state, module.state)
+
+    setupActions(actions, state, module.actions)
+
+    if (module.init) {
+      callbackStack.push(function() {
```
Comment (pspeter3) [inline src/app.js]: To get some bytes back, maybe use bind here?

File: src/app.js [hunk 4]
```diff
@@ -25,107 +25,92 @@ export function app(props) {
...
+          : data && requestRender(copy(withState, data)),
+        withState
+      )
     }
   }
 
-  function set(prop, value, source) {
-    var target = merge(source)
-    target[prop] = value
+  function copy(target, source) {
```
Comment (pspeter3) [inline src/app.js]: Does hyperapp target browsers where this cannot be `Object.assign`?

**Comment by zaceno** [comment]
I thought of a question: Is there any danger in reusing a module? Like what if I say:

```js

const counterModule = {
  state: {value: 0}
  actions: {
    increment: state => ({value: state + 1}),
    decrement: state => ({value: state - 1})
  }
}

app({
  modules: {
    counterA: counterModule,
    counterB: counterModule,
  },
  ...
})
```

does that mean, since I initially defined the two counter modules with the same object reference, that their state will be entangled somehow? Should I instead have written:


```js

const counterModule = _ => {
  state: {value: 0}
  actions: {
    increment: state => ({value: state + 1}),
    decrement: state => ({value: state - 1})
  }
}

app({
  modules: {
    counterA: counterModule(),
    counterB: counterModule(),
  },
  ...
})
```

or is the first option safe? I'm suspecting it is, but I wanted to ask to be sure :)

File: src/app.js (hunk 1 above)
Comment (JorgeBucaran) [reply src/app.js]: @pspeter3 A queue would imply we execute callbacks in the order we found them.

```counter() -> subcounter() -> subsubcounter()```

But instead we execute inside out, the deepest callback first and unwind.

```subsubcounter() -> subcounter() -> counter()```

So, AFAICT, the correct terminology is stack.

File: src/app.js (hunk 4 above)
Comment (JorgeBucaran) [reply src/app.js]: @pspeter3 Yep, so >=IE10.

File: src/app.js (hunk 3 above)
Comment (JorgeBucaran) [reply src/app.js]: @pspeter3 Thanks, I'll see if I can squeeze this any further. 🤓

**Comment by JorgeBucaran** [comment]
@zaceno In your first example, there will be effectively two `{ value: 0 }`: counterA.value and counterB.value. So, it should be fine.

This was a problem during the mixin era since there was no fragmented state and you had to hardcode the namespace inside the mixin definition (bringing with it all the fun of namespace collisions).

With modules, you must specify the namespace to use it: `modules: { counter }`.

File: src/app.js [hunk 5]
```diff
@@ -25,107 +25,92 @@ export function app(props) {
...
-        element.tagName.toLowerCase(),
-        {},
-        map.call(element.childNodes, function(element) {
-          return element.nodeType === 3
-            ? element.nodeValue
-            : hydrate(element, map)
-        })
-      )
-    )
+  function flush(cb) {
```
Comment (johanalkstal) [inline src/app.js]: When does `flush` actually receive a `cb`?

Comment (JorgeBucaran) [reply src/app.js]: To avoid declaring the variable inside the function with `var`. Just a silly technique to save a few bytes here.

File: src/app.js (hunk 1 above)
Comment (pspeter3) [reply src/app.js]: Yeah, makes sense. I think we should document that as people may rely on the execution order.

File: src/app.js [hunk 6]
```diff
@@ -25,107 +25,92 @@ export function app(props) {
...
-          paths.length > 1
-            ? setPath(
-                paths.slice(1),
-                value,
-                source && name in source ? source[name] : {}
-              )
-            : value,
-          source
-        )
+  function merge(target, source, result) {
```
Comment (okwolf) [inline src/app.js]: It looks like this is only ever called with the `target` and `source`, and `result` is always set to `{}`. Will this work just as well?

```js
function merge(target, source) {
  return copy(copy({}, target), source)
}
```

All the tests pass, but for some reason the gzip output is a byte larger 😕

File: src/app.js (hunk 1 above)
Comment (JorgeBucaran) [reply src/app.js]: Execution order within siblings is undetermined though. I think this is not a problem, as it was with mixins because with modules the user, not the author, is responsible to give it a key / scope.

File: src/app.js (hunk 6 above)
Comment (JorgeBucaran) [reply src/app.js]: Good idea, this should work.

File: test/modules.test.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    }
+  }
+
+  app({
+    init(state, actions) {
+      expect(state.foo.value).toBe(0)
+      expect(actions.foo.up().value).toBe(1)
+      done()
+    },
+    modules: { foo }
```
Comment (okwolf) [inline test/modules.test.js]: There should probably be tests for nested modules and `init` on a module.

File: src/app.js [hunk 8]
```diff
@@ -25,107 +25,92 @@ export function app(props) {
...
-    }
-    return result
+  function hydrate(element, map) {
+    return (
+      element &&
+      h(
+        element.tagName.toLowerCase(),
+        {},
+        map.call(element.childNodes, function(element) {
+          return element.nodeType === 3
```
Comment (okwolf) [inline src/app.js]: I'm guessing we can't use `Node.TEXT_NODE` here because uglify won't know to replace it with the constant? I also thought we used to have a comment here about the meaning of this magic number.

Comment (SkaterDad) [reply src/app.js]: Could also copy what Preact does, and check if `element.splitText !== undefined`.   In a conference talk, `developit` mentioned the perf win was substantial, since `nodeType` is a getter function.

https://youtu.be/LY6y3HbDVmg?t=19m53s
https://github.com/developit/preact/blob/master/src/vdom/diff.js#L76

**Comment by zaceno** [comment]
I've been experimenting with this branch, and think there might be a bug in the code that enables the simplified HOA syntax (`app(x)(y)(z)({...})`)

I have two HOA `A` and `B`. When I start my app with `A(B(app))({...})` it works. With `B(A(app))({..})` it works.

However, when I write `app(A)(B)({...})`, `A` fails to do it's thing. And when I write `app(B)(A)({...})` it's `B` that fails.

Am I, as  HOA-author, expected to do anything special in my HOA to enable that "chained" syntax?

File: test/modules.test.js (hunk 7 above)
Comment (JorgeBucaran) [reply test/modules.test.js]: @okwolf Added them both! Please check 🙏

**Comment by JorgeBucaran** [comment]
@okwolf @Swizz @andyrj @zaceno Yes, the correct way to create multiple HOAs that take into account each other's enhancements follow.

```jsx
function B(app) {
  function enhancedApp(props) {
    return app({
      ...props,
      init() {
        console.log("B")
        props.init()
      }
    })
  }

  return props =>
    typeof props === "function" ? props(enhancedApp) : enhancedApp(props)
}

function A(app) {
  function enhancedApp(props) {
    return app({
      ...props,
      init() {
        console.log("A")
        props.init()
      }
    })
  }

  return props =>
    typeof props === "function" ? props(enhancedApp) : enhancedApp(props)
}

app(A)(B)({
  init() {
    console.log("Main")
  }
})
```

**Comment by pspeter3** [comment]
I feel like the HOA abstraction is great but the API feels really awkward when you need to consider all the other higher order apps.

**Comment by JorgeBucaran** [comment]
@pspeter3 What else do you have in mind?

**Comment by JorgeBucaran** [comment]
Thanks everyone! I am ready to merge.

We should address changes to HOA support on a different issue.

/cc @andyrj @dodekeract Comments & feedback please 🙏

**Comment by pspeter3** [comment]
That's exactly what I'm thinking. It seems even less necessary now that there are modules.

**Comment by JorgeBucaran** [comment]
@pspeter3 I hear you. Let's talk about it in https://github.com/hyperapp/hyperapp/issues/413.

File: src/app.js [hunk 9]
```diff
@@ -45,115 +50,86 @@ export function app(props) {
...
+      callbacks.push(function() {
+        module.init(state, actions)
+      })
+    }
+
+    assign(state, module.state)
+
+    initActions(state, actions, module.actions)
+
+    Object.keys(module.modules || {}).map(function(i) {
```
Comment (Swizz) [inline src/app.js]:
```js
for (i in module.modules || {}) {
```

Using it here will save some bytes and give us better perf, I guess.

File: src/app.js [hunk 10]
```diff
@@ -45,115 +50,86 @@ export function app(props) {
...
+      init(module.modules[i], (state[i] = {}), (actions[i] = {}))
+    })
+  }
+
+  function initActions(state, actions, source) {
+    Object.keys(source || {}).map(function(i) {
```
Comment (Swizz) [inline src/app.js]:
```js
for (i in source || {}) {
```

File: src/app.js (hunk 9 above)
Comment (JorgeBucaran) [reply src/app.js]: We will need a function to wrap the expression anyway because we would be accessing a shared variable `i` inside the new action definition `source[i]`. If we could use `let` instead of `var` it would work though.

**Comment by JorgeBucaran** [comment]
Init & modules landed in master! 🎉

File: src/app.js (hunk 3 above)
Comment (Swizz) [reply src/app.js]: `bind` is adding more byte than `function(){ ...` here Oo

**Comment by vdsabev** [comment]
Loving the changes here! And the coverage 😉 
I'll use `modules` for my `router` implementation, and `init` to put some code inside the app definition where I think it'll look better (instead of after).

**Comment by SkaterDad** [comment]
@JorgeBucaran Any idea on when this will get to npm?  Will you wait for the typescript defs to be updated?  I'd love to start refactoring my 0.12.x app now, but will probably wait for the typescript defs for editor tooling support.

**Comment by JorgeBucaran** [comment]
Maybe if we can get @Mytrill @pspeter3 @Swizz to help with the types I could publish 0.15.0 sooner than later? 🤔

**Comment by pspeter3** [comment]
Sure, though I'll likely need some help making sure I understand how modules and init are supposed to work. Perhaps you can modify the TypeScript test and we can make the type definitions work?
